cluster/images/conformance/go-runner/run-conformance.go

```diff
@@ -0,0 +1,341 @@
+}
+
+func main() {
+	env := envWithDefaults(map[string]string{
+		resultsDirEnvKey: defaultResultsDir,
+		skipEnvKey:       defaultSkip,
+		focusEnvKey:      defaultFocus,
+		providerEnvKey:   defaultProvider,
+		parallelEnvKey:   defaultParallel,
+		ginkgoEnvKey:     defaultGinkgoBinary,
+		testBinEnvKey:    defaultTestBinary,
+	})
+
+	if err := run(env); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func run(env Getenver) error {
+	resultsDir := env.Getenv(resultsDirEnvKey)
+	defer saveResults(resultsDir)
+
+	logFilePath := filepath.Join(resultsDir, logFileName)
+	logFile, err := os.Create(logFilePath)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create log file %v", logFilePath)
+	}
+	mw := io.MultiWriter(os.Stdout, logFile)
+	cmd := getCmd(env, mw)
+
+	log.Printf("Running command:\n%v\n", cmdInfo(cmd))
+	err = cmd.Start()
+	if err != nil {
+		return errors.Wrap(err, "starting command")
+	}
+
+	// Handle signals and shutdown process gracefully.
+	go setupSigHandler(cmd.Process.Pid)
+	return errors.Wrap(cmd.Wait(), "running command")
+}
+
+func setupSigHandler(pid int) {
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
+
+	// Block until a signal is received.
+	log.Println("Now listening for interrupts")
+	s := <-c
+	log.Printf("Got signal: %v. Shutting down test process (PID: %v)\n", s, pid)
+	p, err := os.FindProcess(pid)
+	if err != nil {
+		log.Printf("Could not find process %v to shut down.\n", pid)
+		return
+	}
+	if err := p.Signal(s); err != nil {
+		log.Printf("Failed to signal test process to terminate: %v\n", err)
+		return
+	}
+	log.Printf("Signalled process %v to terminate successfully.\n", pid)
+}
+
+// cmdInfo generates a useful look at what the command is for printing/debug.
+func cmdInfo(cmd *exec.Cmd) string {
+	return fmt.Sprintf(
+		`Command env: %v
+Run from directory: %v
+Executable path: %v
+Args (comma-delimited): %v`, cmd.Env, cmd.Dir, cmd.Path, strings.Join(cmd.Args, ","),
+	)
+}
+
+// saveResults will tar the results directory and write the resulting tarball path
+// into the donefile.
+func saveResults(resultsDir string) error {
+	log.Println("Saving results.")
+
+	err := tarDir(resultsDir, filepath.Join(resultsDir, resultsTarballName))
+	if err != nil {
+		return errors.Wrapf(err, "tar directory %v", resultsDir)
+	}
+
+	doneFile := filepath.Join(resultsDir, doneFileName)
+
+	resultsTarball := filepath.Join(resultsDir, resultsTarballName)
+	resultsTarball, err = filepath.Abs(resultsTarball)
+	if err != nil {
+		return errors.Wrapf(err, "failed to find absolute path for %v", resultsTarball)
+	}
+
+	return errors.Wrap(
+		ioutil.WriteFile(doneFile, []byte(resultsTarball), os.FileMode(0777)),
+		"writing donefile",
+	)
+}
+
+// tarDir takes a source and variable writers and walks 'source' writing each file
+// found to the tar writer; the purpose for accepting multiple writers is to allow
+// for multiple outputs (for example a file, or md5 hash)
+func tarDir(dir, outpath string) error {
```

Comment:
should really have a test function for all the messed up path args folks could pass in.

Reply:
Is there some standard you had in mind? The most practical advice I could find on the topic of "check if a path is valid" is to just try and create a tiny file at that location and see if it succeeds.

Since we are creating the file (about 5 lines below your comment) and returning an error there if it fails, that seems like the most practical check already.

Also, I fixed up the error returned via that file creation to be sure to list the path in question.

Finally, since it is going through `os.Create` and not being given to a shell for interpretation, I dont think there is any sort of risk of vulnerability. The default RESULTS_DIR is used in most cases and so this is just using constants that we defined on an unexported method.

Reply:
Write a test function that passes in a series of failure conditions.  TL;DR this package should have unit-tests.

Reply:
I can agree with that general point; I can add some tests (which was the whole point of adding the env stuff: making it testable)

Reply:
Tests added; general topics tested:
 - building the command with a number of different env settings
 - taring a directory results in the expected files/contents
 - taring a non-existant directory errors
 - the env objects behave as expected

It was good that I added these in; I knew I'd have to iterate on it but I did find a bug or two in some of the env parsing. Those have been resolved now.